### PR TITLE
Tensor indexing

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michał Zientkiewicz <mzient@gmail.com>

## Description
- [X] **Bug fix** (*non-breaking change which fixes an issue*)
- [x] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR

Added NumPy-like tensor indexing operator:
- the operator supports up to 32 dimensions with `at`, `lo`, `hi` (and in the future also `step`) arguments.
- the arguments are passed separately for dimensions
- adds `__getitem__` to `DataNode` which instantiates subscript operator
- adds `dali.newaxis` and `dali.newaxis(name)`
- fixes a bug in TensorVector: stale type in sample views when only the  type changes in Resize
- fixes a bug in ExpandDims: incorrenct layout handling for 0D input

##### Affected modules and functionalities:

- DataNode python object
- Operators: added a new operator
- ExpandDims - bug fix
- TensorVector - bug fix

##### Key points relevant for the review:

- Indexing corner cases.

### Tests
- [ ] Existing tests apply
- [ ] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [x] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: SUBSCR.* except SUBSCR.06

**JIRA TASK**: 1617

### CODE REVIEW

File: dali/test/python/test_dali_variable_batch_size.py [hunk 1]
```diff
@@ -1188,12 +1203,13 @@ def pipe(max_batch_size, input_data, device, input_layout=None):
...
     "readers.video_resize",     # readers do do not support variable batch size yet
-
 ]
 
 def test_coverage():
-    methods = module_functions(fn, remove_prefix = "nvidia.dali.fn")
+    methods = module_functions(fn, remove_prefix="nvidia.dali.fn")
     methods += module_functions(dmath, remove_prefix = "nvidia.dali")
-    covered = tested_methods + excluded_methods
+    exclude = "|".join(["(^" + x.replace(".", "\.").replace("*", ".*").replace("?", ".") + "$)" for x in excluded_methods])
```
Comment (JanuszL) [inline dali/test/python/test_dali_variable_batch_size.py:1211]: Do we need this `$` or it is just in case?

Comment (mzient) [reply dali/test/python/test_dali_variable_batch_size.py:1211]: It's formally necessary, even if not needed now - otherwise if we put, say, "paste", here, it would match also "multi_paste" and "bbox_paste".

Comment (JanuszL) [reply dali/test/python/test_dali_variable_batch_size.py:1211]: Rather "paste_x" and "paste_y". `^` should make sure of that. Anyway, just asking.

File: dali/pipeline/data/tensor_vector.cc [hunk 2]
```diff
@@ -461,6 +461,8 @@ void TensorVector<Backend>::update_view(int idx) {
     tensors_[idx]->ShareData(std::shared_ptr<void>(ptr, ViewRefDeleter{&views_count_}),
                              volume(tl_->tensor_shape(idx)) * tl_->type().size(), shape,
                              tl_->type());
+  } else if (IsValidType(tl_->type())) {
+    tensors_[idx]->set_type(tl_->type());
```
Comment (mzient) [inline dali/pipeline/data/tensor_vector.cc:465]: This kept stale type in the views if the pointer or size didn't change.

File: dali/operators/generic/slice/subscript.h [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+  void Init(const OpSpec &spec, int i) {
+    at.Init(spec, make_string("at_", i));
+    lo.Init(spec, make_string("lo_", i));
+    hi.Init(spec, make_string("hi_", i));
+    step.Init(spec, make_string("step_", i));
+
+    if (at.IsDefined()) {
+      if (lo.IsDefined() || hi.IsDefined()) {
+        DALI_FAIL(make_string("The subscriptg for dimension ", i,
+                              " is specified both as an index and as a range."));
```
Comment (JanuszL) [inline dali/operators/generic/slice/subscript.h:103]:
```suggestion
        DALI_FAIL(make_string("The subscriptg for dimension ", i,
                              " should not be specified both as an index and as a range."));
```

File: dali/operators/generic/slice/subscript.h [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+      }
+      if (step.IsDefined()) {
+        DALI_FAIL(make_string("The subscriptg for dimension ", i,
+                              " is specified as an index - it cannot have a step."));
```
Comment (JanuszL) [inline dali/operators/generic/slice/subscript.h:107]:
```suggestion
                              " should not be specified as an index and have have a step."));
```

File: dali/operators/generic/slice/subscript.h [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+  void InitArgs() {
+    subscripts_.resize(kMaxSubscripts);
+    int num_subscripts = 0;
+    for (int i = 0; i < kMaxSubscripts; i++) {
+      SubscriptInfo &s = subscripts_[i];
+      s.Init(spec_, i);
+
+
+      if (s.IsDefined()) {
+        num_subscripts = i + 1;
```
Comment (JanuszL) [inline dali/operators/generic/slice/subscript.h:153]:
```suggestion
        max_subscripts = i + 1;
```
As it is not a number because some subscripts may be not defined.

File: dali/operators/generic/slice/subscript.h [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+    outputs[0].type = input.type();
+    GetRanges(ws, input_shape);
+    ProcessShapes(outputs[0].shape, input_shape);
+    return true;
+  }
+
+  void GetRanges(const workspace_t<Backend> &ws, const TensorListShape<> &in_shape) {
+    int nsub = subscripts_.size();
+    int ndim = in_shape.sample_dim();
+    DALI_ENFORCE(ndim >= nsub, "Too many indices.");
```
Comment (JanuszL) [inline dali/operators/generic/slice/subscript.h:173]:
```suggestion
    DALI_ENFORCE(ndim >= nsub, make_string("Too many indices. Received: ", nsub, ", while input has only " ndim, " dimensions.");
```

File: dali/operators/generic/slice/subscript.h [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+
+  void GetRange(SubscriptInfo &s, int d, const TensorListShape<> &in_shape) {
+    int nsamples = in_shape.num_samples();
+
+    if (s.IsIndex()) {
+      for (int i = 0; i < nsamples; i++) {
+        int64_t in_extent = in_shape.tensor_shape_span(i)[d];
+        int64_t at = s.at.values[i];
+        int64_t idx = at < 0 ? in_extent + at : at;
+        DALI_ENFORCE(idx >= 0 && idx < in_extent, make_string("Index ", idx, " is out of range "
```
Comment (JanuszL) [inline dali/operators/generic/slice/subscript.h:197]:
```suggestion
        DALI_ENFORCE(idx >= 0 && idx < in_extent, make_string("Index ", idx, "(", at, ") is out of range "
```

File: dali/operators/generic/slice/subscript.h [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+        shape_.tensor_shape_span(i)[d] = 1;
+      }
+    }
+    if (s.IsRange()) {
+      for (int i = 0; i < nsamples; i++) {
+        int64_t in_extent = in_shape.tensor_shape_span(i)[d];
+        int64_t lo = s.lo.IsDefined() ? s.lo.values[i] : 0;
+        int64_t hi = s.hi.IsDefined() ? s.hi.values[i] : in_extent;
+        int64_t step = s.step.IsDefined() ? s.step.values[i] : 1;
+        DALI_ENFORCE(step == 1, "Indexing with non-unit step is not implemented");
```
Comment (JanuszL) [inline dali/operators/generic/slice/subscript.h:211]:
```suggestion
        DALI_ENFORCE(abs(step) == 1, "Indexing with non-unit step is not implemented");
```
I think negative step is allowed too (based on L220).

File: dali/operators/generic/slice/subscript.h [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+
+    out_dim_map_.clear();
+    dim_groups_.clear();
+
+    int group_start = 0;
+    int d = 0;
+    for (; d < nsub; d++) {
+      if (subscripts_[d].IsDefined()) {
+        if (d != group_start)
+          dim_groups_.push_back({ group_start, d - group_start });
```
Comment (JanuszL) [inline dali/operators/generic/slice/subscript.h:241]: This contains groups of dims where subscripts_ is not defined, right?

File: dali/operators/generic/slice/subscript.h [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+    if (in_dims != group_start)
+      dim_groups_.push_back({ group_start, in_dims - group_start });
+
+    collapse_dims(simplified_in_shape_, in_shape, dim_groups_);
+    collapse_dims(simplified_out_shape_, shape_, dim_groups_);
+    collapse_dims(simplified_anchor_, lo_, dim_groups_);
+
+    out_shape.resize(in_shape.num_samples(), out_dim_map_.size());
+    for (int i = 0; i < out_shape.num_samples(); i++) {
+      auto out_sample = out_shape.tensor_shape_span(i);
```
Comment (JanuszL) [inline dali/operators/generic/slice/subscript.h:262]:
```suggestion
      auto out_sample_shape = out_shape.tensor_shape_span(i);
```

File: dali/operators/generic/slice/subscript.h [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+
+  bool CanInferOutputs() const override { return true; }
+
+  using Operator<Backend>::RunImpl;
+  void RunImpl(workspace_t<Backend> &ws) override {
+    const auto &input = ws.template InputRef<Backend>(0);
+    auto &output = ws.template OutputRef<Backend>(0);
+    output.SetLayout(GetOutputLayout(input.GetLayout()));
+    VALUE_SWITCH(simplified_in_shape_.sample_dim(), ndim,
+      (1,  2,  3,  4,  5/*,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16*/),
```
Comment (JanuszL) [inline dali/operators/generic/slice/subscript.h:288]: You can add a comment why it is commented out.

File: dali/python/nvidia/dali/data_node.py [hunk 12]
```diff
@@ -122,6 +139,63 @@ def __bool__(self):
...
+        new_axis_names = []
+
+        # returns True if this index adds a new output dimension
+        def process_index(idx, dim):
+            if idx is None:
+                idxs.append((None, None, None, None))
+                return True
+            elif isinstance(idx, slice):
+                if idx.step is not None and idx.step != 1:
+                    raise NotImplementedError("Slicing with non-unit step is not implemented.")
```
Comment (JanuszL) [inline dali/python/nvidia/dali/data_node.py:154]: Do we need it here too or the native side is sufficient?

Comment (JanuszL) [reply dali/python/nvidia/dali/data_node.py:154]: And how about negative step (`-1`)?

File: dali/test/python/test_operator_subscript.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    src = fn.external_source(lambda: data, device=device)
+    pipe.set_outputs(fn.tensor_subscript(src, **args))
+    with assert_raises(RuntimeError):
+        pipe.build()
+
+def test_inconsistent_args():
+    for device in ["cpu", "gpu"]:
+        for args in [
+                { "lo_0":0, "at_0":0 },
+                { "at_0":0, "step_0":1 },
```
Comment (JanuszL) [inline dali/test/python/test_operator_subscript.py:125]: Can you also check step > 1?

File: dali/operators/generic/slice/subscript.h [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+  vector<int64_t> values;
+  int64_t scalar_value = 0;
+
+  void Load(const ArgumentWorkspace &ws, int batch_size) {
+    if (src == ArgSource::Input) {
+      const TensorVector<CPUBackend> &inp = ws.ArgumentInput(name);
+      const auto &shape = inp.shape();
+
+      int n = batch_size;
+      DALI_ENFORCE(shape.num_samples() == n, make_string(
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:72]: This is called at every batch. Currently the error condition is impossible to trigger (it's for the future and fully variable batch size).

File: dali/operators/generic/slice/subscript.h (hunk 3 above)
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:103]: must not ;)

File: dali/operators/generic/slice/subscript.h (hunk 5 above)
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:153]: Mixed feelings here. It's used as the number (see `subscripts_.resize(num_subscripts);`). I would gladly use input ndim here _if I had it_ - but it's not available at construction.

File: dali/operators/generic/slice/subscript.h (hunk 8 above)
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:211]: No, it's not. Line 220 is future-proof, but the step is ignored, since the kernel doesn't support it yet.

File: dali/operators/generic/slice/subscript.h (hunk 9 above)
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:241]: I can rename to `collapsed_dims_` or sth.

File: dali/operators/generic/slice/subscript.h (hunk 11 above)
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:288]: Mhh... It shouldn't be.  It's commented to save time in local builds (and it's _very_ slow to build).  I hope I remember to uncomment it in the final commit.

File: dali/python/nvidia/dali/data_node.py (hunk 12 above)
Comment (mzient) [reply dali/python/nvidia/dali/data_node.py:154]: Not supported yet.

File: dali/test/python/test_operator_subscript.py (hunk 13 above)
Comment (mzient) [reply dali/test/python/test_operator_subscript.py:125]: Well.. I could, but on the other hand, that's "not implemented", not "user error" . I can add it, I guess.

File: dali/operators/generic/slice/subscript.h (hunk 8 above)
Comment (JanuszL) [reply dali/operators/generic/slice/subscript.h:211]: So maybe comment out L200 not to confuse the reader?

File: dali/operators/generic/slice/subscript.h (hunk 9 above)
Comment (JanuszL) [reply dali/operators/generic/slice/subscript.h:241]: Sounds better

File: dali/test/python/test_operator_subscript.py (hunk 13 above)
Comment (JanuszL) [reply dali/test/python/test_operator_subscript.py:125]: Just wanted to make sure that we error as we should.

File: dali/operators/generic/slice/subscript.h (hunk 8 above)
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:211]: I'd rather keep it and add strides sooner than later.

Comment (JanuszL) [reply dali/operators/generic/slice/subscript.h:211]: Well, dead code is a bit misleading. I'm not asking to remove but comment out for now.

File: dali/operators/generic/slice/subscript.h (hunk 7 above)
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:197]: I'll just use `at` here - using two values for positive indices would look weird and for negative indices we could assume that the users know what they're doing.

Comment (mzient) [reply dali/operators/generic/slice/subscript.h:197]: ...and I'll replace `DALI_ENFORCE` with plain `if` and `DALI_FAIL` to avoid putting the confusing condition in the error message.

File: dali/python/nvidia/dali/data_node.py [hunk 15]
```diff
@@ -122,6 +139,63 @@ def __bool__(self):
                 " represent computations in DALI Pipeline see the \"Mathematical Expressions\""
                 " section of DALI documentation."))
 
+    def __getitem__(self, val):
```
Comment (klecki) [inline dali/python/nvidia/dali/data_node.py:142]: I think that with `__getitem__` defined DataNode is now iterable, you should be able to write:

```
for i in DataNode():
    print("I have no idea what is happening anymore")  
```
and it will do some looping. Unless what we did to prohibit `if DataNode():` still applies here.

File: dali/python/nvidia/dali/data_node.py [hunk 16]
```diff
@@ -122,6 +139,63 @@ def __bool__(self):
...
+        idxs = []
+        new_axes = []
+        new_axis_names = []
+
+        # returns True if this index adds a new output dimension
+        def process_index(idx, dim):
+            if idx is None:
```
Comment (klecki) [inline dali/python/nvidia/dali/data_node.py:149]: How/when can we get `None` here?

File: dali/python/nvidia/dali/data_node.py [hunk 17]
```diff
@@ -122,6 +139,63 @@ def __bool__(self):
                 " represent computations in DALI Pipeline see the \"Mathematical Expressions\""
                 " section of DALI documentation."))
 
+    def __getitem__(self, val):
+        idxs = []
+        new_axes = []
+        new_axis_names = []
+
+        # returns True if this index adds a new output dimension
+        def process_index(idx, dim):
```
Comment (klecki) [inline dali/python/nvidia/dali/data_node.py:148]: I think we need a lot of type checking in this function. This is Python, so we can and will get stuff like DataNode()[DataNode()], DataNode()[np.array(...)], DataNode()["duck"].

I think we are limiting to integers now, right?

File: dali/python/nvidia/dali/data_node.py [hunk 18]
```diff
@@ -122,6 +139,63 @@ def __bool__(self):
...
+                if idx.step is not None and idx.step != 1:
+                    raise NotImplementedError("Slicing with non-unit step is not implemented.")
+                idxs.append((None, idx.start, idx.stop, None))
+                return True
+            elif isinstance(idx, _NewAxis):
+                new_axes.append(dim)
+                if idx.name is not None:
+                    new_axis_names.append(idx.name)
+                return True
+            else:
```
Comment (klecki) [inline dali/python/nvidia/dali/data_node.py:162]: I think this should be case for integers, and we also can handle `...` (Elipsis).

**Review by klecki — COMMENTED** [review]
Comments to first file.

File: dali/test/python/test_operator_subscript.py (hunk 13 above)
Comment (mzient) [reply dali/test/python/test_operator_subscript.py:127]: Done in a separate test.

File: dali/python/nvidia/dali/data_node.py (hunk 18 above)
Comment (mzient) [reply dali/python/nvidia/dali/data_node.py:162]: We don't handle ellipsis... yet. It would actually be quite hard to add.

File: dali/python/nvidia/dali/data_node.py (hunk 17 above)
Comment (mzient) [reply dali/python/nvidia/dali/data_node.py:148]: Now, we're not and we do support datanodes and slices with datanodes and a mix of datanodes and integers.

File: dali/python/nvidia/dali/data_node.py (hunk 16 above)
Comment (mzient) [reply dali/python/nvidia/dali/data_node.py:149]: I must recheck what happens with `:`.

File: dali/python/nvidia/dali/data_node.py (hunk 15 above)
Comment (mzient) [reply dali/python/nvidia/dali/data_node.py:145]: Fixed - `__iter__` now throws a TypeError.

File: dali/python/nvidia/dali/data_node.py (hunk 18 above)
Comment (klecki) [reply dali/python/nvidia/dali/data_node.py:165]: ack, that is just me starting with this file, didn't think about the use of ArgumentInputs.

File: dali/python/nvidia/dali/data_node.py (hunk 17 above)
Comment (klecki) [reply dali/python/nvidia/dali/data_node.py:151]: Ack. Didn't think about Argument Inputs. Just wonder how the errors would look if we use a float scalar here for example.

File: dali/operators/generic/slice/subscript.cc [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+  auto &output = ws.template OutputRef<CPUBackend>(0);
+  int N = input.ntensor();
+  using T = kernels::type_of_size<element_size>;
+  ThreadPool &tp = ws.GetThreadPool();
+
+  kernels::SliceCPU<T, T, ndim> K;
+  TensorView<StorageCPU, const T, ndim> tv_in;
+  TensorView<StorageCPU, T, ndim> tv_out;
+
+  kernels::KernelContext ctx;
```
Comment (klecki) [inline dali/operators/generic/slice/subscript.cc:81]: Is having one ctx for all kernels here ok?

**Review by klecki — COMMENTED** [review]
Comments to subscript.{cu, cc, h}.

File: dali/pipeline/data/tensor_vector.cc (hunk 2 above)
Comment (klecki) [reply dali/pipeline/data/tensor_vector.cc:465]: We have Reset above, so it happened only with empty TV that got a type?

File: dali/test/python/test_dali_cpu_only.py [hunk 20]
```diff
@@ -99,9 +100,9 @@ def check_mixed_op_bad_device(device_id):
     pipe.set_outputs(decoded)
     assert_raises(RuntimeError, pipe.build)
 
-def test_mixed_op_bad_device():
-    for device_id in [None, 0]:
-        yield check_bad_device, device_id
+#def test_mixed_op_bad_device():
```
Comment (klecki) [inline dali/test/python/test_dali_cpu_only.py:103]: why is this commented?

File: dali/python/nvidia/dali/_utils/hacks.py [hunk 21]
```diff
@@ -0,0 +1,49 @@
+# Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (klecki) [inline dali/python/nvidia/dali/_utils/hacks.py:1]: Do we really want to name it `hacks`?

File: dali/python/nvidia/dali/_utils/hacks.py [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+        return False
+    return _original_check(iterable, instance)
+
+def _hook_iterable_check():
+    global _original_check
+    global _not_iterable
+    if _original_check is not None:
+        return
+    _original_check = type(collections.abc.Iterable).__instancecheck__
+    type(collections.abc.Iterable).__instancecheck__ = _check_iterable
```
Comment (klecki) [inline dali/python/nvidia/dali/_utils/hacks.py:35]: After calling it twice,
```
_original_check = _check_iterable
type(collections.abc.Iterable).__instancecheck__ = _check_iterable
```


Calling isinstance(x, Iterable) will call `_check_iterable` which calls `_original_check` (set to `_check_iterable`, and we have infinite recursion I think.

File: dali/python/nvidia/dali/_utils/hacks.py [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+    if len(_not_iterable) == 0:
+        _not_iterable = (_NotIterable,)
+
+def not_iterable(cls, add_iter=True):
+    """Makes an object non-iterable by raising a TypeError in add_iter and suppressing
```
Comment (klecki) [inline dali/python/nvidia/dali/_utils/hacks.py:40]:
```suggestion
    """Makes an object non-iterable by raising a TypeError in `__iter__` and suppressing
```

**Review by klecki — COMMENTED** [review]
Comments to the hack.

I need to look at the test on Monday.

File: dali/python/nvidia/dali/_utils/hacks.py (hunk 21 above)
Comment (mzient) [reply dali/python/nvidia/dali/_utils/hacks.py:1]: Well, what else is that?

Comment (JanuszL) [reply dali/python/nvidia/dali/_utils/hacks.py:1]: Workarounds?

File: dali/operators/generic/slice/subscript.cc (hunk 19 above)
Comment (mzient) [reply dali/operators/generic/slice/subscript.cc:81]: Yeah, it's completely stateless. As you can see, I'm not running setup nor providing a scratchpad.

File: dali/operators/generic/slice/subscript.h [hunk 24]
```diff
@@ -0,0 +1,337 @@
...
+    outputs[0].type = input.type();
+    GetRanges(ws, input_shape);
+    ProcessShapes(outputs[0].shape, input_shape);
+    return true;
+  }
+
+  void GetRanges(const workspace_t<Backend> &ws, const TensorListShape<> &in_shape) {
+    int nsub = subscripts_.size();
+    int ndim = in_shape.sample_dim();
+    DALI_ENFORCE(ndim >= nsub_declared_,
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:180]: I don't need to process the trailing (declared) dimensions if they are full-range. I need to check that `ndim >= nsub_declared_` _somewhere_ and that somewhere is here. I could move the test (to setupimpl?).

File: dali/operators/generic/slice/subscript.h [hunk 25]
```diff
@@ -0,0 +1,337 @@
...
+
+    if (s.IsIndex()) {
+      for (int i = 0; i < nsamples; i++) {
+        int64_t in_extent = in_shape.tensor_shape_span(i)[d];
+        int64_t at = s.at.values[i];
+        int64_t idx = at < 0 ? in_extent + at : at;
+        if (idx < 0 || idx >= in_extent)
+          DALI_FAIL(make_string("Index ", at, " is out of range "
+            "for axis ", d, " of length ", in_extent, "\n"
+            "Detected while processing sample #", i, " of shape (", in_shape[i], ")"));
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:209]: No. For 1D it's just a number, for higher dims it's like `480x640x3`. It looked weird for 1D without parenthesis (`... of shape 3`).

File: dali/operators/generic/slice/subscript.h [hunk 26]
```diff
@@ -0,0 +1,337 @@
...
+    for (; d < in_dims; d++) {
+      out_dim_map_.push_back(d);
+    }
+
+    if (in_dims != group_start)
+      collapsed_dims_.push_back({ group_start, in_dims - group_start });
+
+    collapse_dims(simplified_in_shape_, in_shape, collapsed_dims_);
+    collapse_dims(simplified_out_shape_, shape_, collapsed_dims_);
+    collapse_dims(simplified_anchor_, lo_, collapsed_dims_);
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:268]: It should be, I guess.

File: dali/operators/generic/slice/subscript.h [hunk 27]
```diff
@@ -0,0 +1,337 @@
...
+    auto &output = ws.template OutputRef<Backend>(0);
+    output.SetLayout(GetOutputLayout(input.GetLayout()));
+    VALUE_SWITCH(simplified_in_shape_.sample_dim(), ndim,
+      (1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16),
+      (VALUE_SWITCH(input.type().size(), element_size, (1, 2, 4, 8),
+        (RunTyped<ndim, element_size>(ws);),
+        (DALI_FAIL(make_string("Unsupported input type: ", input.type().id()));))),
+      (DALI_FAIL("Subscript too complex.\n"
+        "The subscript operator supports up to 32 total and up to  16 non-collapsible dimensions.\n"
+        "Adjacent dimensions from which no index or slice is taken can be collapsed.");)
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:304]: That's merely an optimization. We can detect it in a follow-up, when we add support strided ranges, for example.

File: dali/pipeline/data/tensor_vector.cc (hunk 2 above)
Comment (mzient) [reply dali/pipeline/data/tensor_vector.cc:465]: It happend with something that had non-empty shape and type but 0 volume.

File: dali/python/nvidia/dali/_utils/hacks.py (hunk 21 above)
Comment (mzient) [reply dali/python/nvidia/dali/_utils/hacks.py:1]: You work around a bug or a lack of feature. Here we try to get around _a feature_. I'd call that hacking. And c'mon, I've seen this name in several projects before.

File: dali/operators/generic/slice/subscript.h (hunk 8 above)
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:221]: I'll add a comment about it, but I prefer to keep the code in zombie state, but at least compiling.

File: dali/python/nvidia/dali/_utils/hacks.py (hunk 22 above)
Comment (mzient) [reply dali/python/nvidia/dali/_utils/hacks.py:35]: There's a check for that 2 lines before.

File: dali/test/python/test_dali_cpu_only.py (hunk 20 above)
Comment (mzient) [reply dali/test/python/test_dali_cpu_only.py:103]: because.... I have a GPU and it failed.

File: dali/test/python/test_pipeline.py [hunk 28]
```diff
@@ -1609,6 +1609,21 @@ def my_pipe():
 
     p = my_pipe(device_id=0, seed=1234, num_threads=3, set_affinity=True, py_num_workers=3)
 
+def test_not_iterable():
+    import nvidia.dali._utils.hacks as hacks
```
Comment (JanuszL) [inline dali/test/python/test_pipeline.py:1613]: Why hacks? Workarounds?

File: dali/test/python/test_operator_subscript.py [hunk 29]
```diff
@@ -0,0 +1,190 @@
...
+        inp, cpu, gpu = pipe.run()
+        for i in range(len(inp)):
+            x = inp.at(i)
+            ref = x[lo_idxs[j]:, :hi_idxs[k]]
+            j = (j + 1) % len(lo_idxs)
+            k = (k + 1) % len(hi_idxs)
+            assert np.array_equal(ref, cpu.at(i))
+            assert np.array_equal(ref, gpu.as_cpu().at(i))
+
+def test_constant_ranges():
```
Comment (JanuszL) [inline dali/test/python/test_operator_subscript.py:105]: I think it repeats L56 and hides it.

File: dali/test/python/test_operator_subscript.py [hunk 30]
```diff
@@ -0,0 +1,190 @@
...
+
+def _test_invalid_args(device, args, run):
+    data = [np.uint8([[1,2,3]]),np.uint8([[1,2]])]
+    pipe = Pipeline(1, 1, 0)
+    src = fn.external_source(lambda: data, device=device)
+    pipe.set_outputs(fn.tensor_subscript(src, **args))
+    with assert_raises(RuntimeError):
+        pipe.build()
+        if run:
+            pipe.run()
```
Comment (JanuszL) [inline dali/test/python/test_operator_subscript.py:121]: Can we tell when it should error, build or runtime?
Maybe we can test if the error is raised at the proper level?

File: dali/python/nvidia/dali/data_node.py [hunk 31]
```diff
@@ -122,6 +142,73 @@ def __bool__(self):
...
 
+    def __getitem__(self, val):
+        idxs = []
+        new_axes = []
+        new_axis_names = []
+
+        # returns True if this index adds a new output dimension
+        def process_index(idx, dim):
+            if idx is None:
+                print("DUPA!")
```
Comment (JanuszL) [inline dali/python/nvidia/dali/data_node.py:153]: Please remove.

File: dali/python/nvidia/dali/data_node.py [hunk 32]
```diff
@@ -122,6 +142,73 @@ def __bool__(self):
...
+        for i, (at, lo, hi, step) in enumerate(idxs):
+            if at   is not None: slice_args["at_%i"%i] = at
+            if lo   is not None: slice_args["lo_%i"%i] = lo
+            if hi   is not None: slice_args["hi_%i"%i] = hi
+            if step is not None: slice_args["step_%i"%i] = step
+
+        import nvidia.dali.fn
+        if len(slice_args) == 0:
+            if len(new_axes) == 0 or new_axes[-1] < len(idxs):
+                print("Adding dim check for ", len(idxs))
```
Comment (JanuszL) [inline dali/python/nvidia/dali/data_node.py:199]: Do we need these prints?

File: dali/python/nvidia/dali/_utils/hacks.py (hunk 21 above)
Comment (JanuszL) [reply dali/python/nvidia/dali/_utils/hacks.py:1]: I still have a very mixed feelings about this....

File: dali/operators/generic/slice/subscript.h [hunk 33]
```diff
@@ -0,0 +1,358 @@
...
+struct SubscriptInfo {
+  void Init(const OpSpec &spec, int i) {
+    at.Init(spec, make_string("at_", i));
+    lo.Init(spec, make_string("lo_", i));
+    hi.Init(spec, make_string("hi_", i));
+    step.Init(spec, make_string("step_", i));
+
+    if (at.IsDefined()) {
+      if (lo.IsDefined() || hi.IsDefined()) {
+        DALI_FAIL(make_string("The subscriptg for dimension ", i,
```
Comment (JanuszL) [inline dali/operators/generic/slice/subscript.h:102]:
```suggestion
        DALI_FAIL(make_string("The subscript for dimension ", i,
``?

File: dali/operators/generic/slice/subscript.h [hunk 34]
```diff
@@ -0,0 +1,358 @@
...
+                              " must not be specified both as an index and as a range."));
+      }
+      if (step.IsDefined()) {
+        DALI_FAIL(make_string("The subscriptg for dimension ", i,
```
Comment (JanuszL) [inline dali/operators/generic/slice/subscript.h:106]:
```suggestion
        DALI_FAIL(make_string("The subscript for dimension ", i,
```
?

File: dali/operators/generic/slice/subscript.h (hunk 33 above)
Comment (mzient) [reply dali/operators/generic/slice/subscript.h:102]: Typo.

File: dali/python/nvidia/dali/data_node.py (hunk 32 above)
Comment (mzient) [reply dali/python/nvidia/dali/data_node.py:199]: Surely not. These are remnants from initial testing - I wanted to be 100% sure that the code follows the intended path and that the right checks fire when expected.

File: dali/test/python/test_operator_subscript.py (hunk 29 above)
Comment (mzient) [reply dali/test/python/test_operator_subscript.py:105]: Some copy paste... I'll take a look.

File: dali/python/nvidia/dali/data_node.py (hunk 31 above)
Comment (mzient) [reply dali/python/nvidia/dali/data_node.py]: Done ;)

File: dali/test/python/test_operator_subscript.py (hunk 30 above)
Comment (mzient) [reply dali/test/python/test_operator_subscript.py]: Out-of-range indices fail at run-time, inconsistent arguments fail at build time. What we don't detect is that a pipeline expected to fail at run-time actually failed at build-time - but we check the message.

**Review by JanuszL — APPROVED** [review]

**Review by klecki — APPROVED** [review]

**Review by klecki — APPROVED** [review]
